### PR TITLE
Flow Test

### PR DESCRIPTION
**Pull request by joe-gre**
Adds a generic test used to test flow output. This is meant to be called in flow tests with input and output objects after setup is complete. Also adds tests for this test.

### CODE REVIEW

**Comment by joe-gre** [comment]
> We need to make it more generic.
> 
> You are making it too tightly coupled with RabbitMQ.
> 
> What if I need to pass a connection to Redis? To Postgres db? I will have to change to library every time.
> 
> We need to provide an API that would allow to do these things without changing the library.

The point of this module is to prevent our Node-RED flow tests from having to duplicate logic. It is not meant to be a generic container/resource manager. Basically we don't want to have to have the RabbitMQ (and other communication mechanism) specific logic in flow-test.ts duplicated in every flow repo (sounds like a nightmare to maintain).

If we want to make test output in Redis or Postgres, it is best to make the change in this library because this change could involve some fairly complex logic that we don't want to have to copy/paste and maintain separately. Changing the library is simple, changing many many tests in different repos with the needed changes is not.

If you want to pass input in or consume output from something other than AMQP you can extend Input or Output and provide the logic in flow-test.ts (this file should probably be broken up in the future). I did this to allow HTTP input (output still needs to be done).

File: src/flow/flow-test.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        resolve: undefined,
+        reject: undefined
+    };
+
+    const consumerPromise = new Promise<void>((resolve, reject) => {
+        res.resolve = resolve;
+        res.reject = reject;
+    });
+
+    const timeout = setTimeout(async () => {
```
Comment (ziflex) [inline src/flow/flow-test.ts]: I don't think you need to make this function async.

File: src/flow/flow-test.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                    throw new Error(`Output of queue ${outQueue} is null`);
+                }
+
+                if (!shouldReceive) {
+                    throw new Error(`Queue ${outQueue} received a message when it shouldn't have`);
+                }
+
+                const out = JSON.parse(msg.content.toString());
+
+                expect(out).to.deep.equal(expectedOutput, `Output of queue ${outQueue} does not match expected output`);
```
Comment (ziflex) [inline src/flow/flow-test.ts]: ``chai`` is a dev dependency.
You should either move it to prod dependency list or use an assertion library from [the standard library](https://nodejs.org/dist/latest-v12.x/docs/api/assert.html).

File: src/flow/flow-test.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                res.resolve();
```
Comment (ziflex) [inline src/flow/flow-test.ts]: What if it's been resolved in your timeout function? 
There must be a check that handles it.

File: package.json [hunk 4]
```diff
@@ -52,6 +51,7 @@
     "dependencies": {
         "amqplib": "^0.5.5",
         "axios": "^0.19.2",
+        "chai": "^4.2.0",
```
Comment (ziflex) [inline package.json]: I don't think it's a good idea to use it in prod.
Just use ``assert`` module. It's a best practice.

Comment (joe-gre) [reply package.json]: Why? Shouldn't we be consistent with how we test to keep things simple? This is a module to aid testing.

Comment (ziflex) [reply package.json]: True to that.

**Pull request merged**